---
title: Single paternal glucocorticoid receptor challenge reveals circRNA as candidate
  in sperm-RNA mediated inheritance
output:
  html_document: default
  html_notebook: default
---





```{r, include=FALSE}
library(readr)
library(data.table)
library(ggplot2)
library(SingleCellExperiment)
library(SC3)
library(scater)
library(cowplot)
library(scmap)
library(biomaRt)


```


This report compile all the results and computational analyses peformed to analyse RNA-seq comming from 2-cell embryos. 


# Input data and metadata analysis


```{r, message=FALSE}
study_info <- data.table(read_delim("./data/study_info.txt", "\t", escape_double = FALSE, trim_ws = TRUE))

sample_table <- data.table(read_csv("./data/sample_table.csv"))

metadata <- merge(sample_table, study_info, by.x="sample_donor_id", by.y="sanger sample ID")



```

 We processed a total number of `r nrow( metadata[ Relevant=="yes" & !is.na(treatment), ])` cells

```{r}

 metadata[ Relevant=="yes" & !is.na(treatment), .N , by="treatment" ]

```


Using featureCounts we generated a count tables for each cell, which include also ERRCs.


```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=6}


TOTAL_gene_count <- data.table(read_delim("data/TOTAL.gene_count.txt", delim="\t", comment="# "))



n <- ncol(TOTAL_gene_count)

gene_lenght <- TOTAL_gene_count[, c("Geneid", "Length")]

TOTAL_gene_count[, c(seq(2,6)):=NULL ]


TOTAL_gene_count_melted <- data.table::melt(TOTAL_gene_count, value.name = "count", variable.name = "cell")

TOTAL_gene_count_melted[, cell:=substring(cell, 12, 22)]

ERCC_counts <- TOTAL_gene_count_melted[startsWith(Geneid ,"ERCC"), .(count=sum(count), type="ERCC"), by="cell"]


```


Extracting gene names and identifing mitocondrial genes


```{r}


ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl")


total_genes <-  data.table(getBM(attributes=c("ensembl_gene_id", "mgi_symbol"),filters = 'ensembl_gene_id', values = vapply(strsplit(TOTAL_gene_count$Geneid, ".", fixed = TRUE), "[", "", 1) , mart = ensembl))


mt_genes <- total_genes[grepl('^mt-', mgi_symbol) == TRUE , ensembl_gene_id]
```


Sparating ECRR counts, mitochondrial counts and gene counts

```{r}

ERCC_counts <- TOTAL_gene_count_melted[startsWith(Geneid ,"ERCC"), .(count=sum(count), type="ERCC"), by="cell"]

total_genes_melted <- rep(vapply(strsplit(TOTAL_gene_count$Geneid, ".", fixed = TRUE), "[", "", 1), ncol(TOTAL_gene_count) - 1)


mit_gene_counts <- TOTAL_gene_count_melted[which(total_genes_melted %in% mt_genes), sum(count), by=cell]

colnames(mit_gene_counts) <- c("cell", "count")
mit_gene_counts[, type:="Mitochondrial"]



Gene_counts <- TOTAL_gene_count_melted[which(! total_genes_melted %in% mt_genes & ! startsWith(total_genes_melted ,"ERCC") ), sum(count), by=cell]
Gene_counts[,type:="Gene"]
colnames(Gene_counts) <- c("cell", "count", "type" )

```



```{r}



ERCC_mit_Gene_counts <- rbind(ERCC_counts, Gene_counts, mit_gene_counts)


ERCC_mit_Gene_counts[, `:=`(total_count=sum(count)   ), by=cell]
ERCC_mit_Gene_counts[type=="ERCC" , `:=`(ERCC_percentage = count*100/total_count    ), by=cell]


```


Visualizing the read count content for each cell used for the analysis.


```{r}

mit_percent <- ERCC_mit_Gene_counts[type=="Mitochondrial", .(cell, mit_percent=count*100/total_count)]
ERCC_percent <- ERCC_mit_Gene_counts[type=="ERCC", .(cell, ERCC_percent=count*100/total_count)]



#sorted_cells = unique(ERCC_mit_Gene_counts[ order(total_count), cell])

sorted_cells = mit_percent[order(mit_percent), cell]
ERCC_mit_Gene_counts[, cell:=factor(cell, levels= sorted_cells )]

ggplot(ERCC_mit_Gene_counts) +
  geom_bar(aes(x=cell, y=count, fill=type), stat="identity", position = "stack") +
  theme_bw()+
  theme( axis.text.x=element_blank())


ggplot(ERCC_mit_Gene_counts) +
  geom_bar(aes(x=cell, y=count, fill=type), stat="identity", position = "fill") +
  geom_hline(yintercept = 0.15, color="black", linetype="dashed") +
  theme_bw()+
  theme( axis.text.x=element_blank())


ggplot(ERCC_mit_Gene_counts) +
  geom_bar(aes(x=cell, y=count, fill=type, group=(ERCC_percentage>=10)), stat="identity", position = "fill") +
  
  theme_bw()+
  theme( axis.text.x=element_blank())



```


Based on the previous plot, we defined ad oc filters for the cells

```{r}
filtered_out <- c(as.character(mit_percent[mit_percent>15, cell]), as.character( ERCC_percent[ERCC_percent>10, cell]), as.character(ERCC_mit_Gene_counts[total_count<5e5,cell]))
filtered_out <- unique(filtered_out)

```


Here are the results of such filters

```{r}
sorted_cells = unique(ERCC_mit_Gene_counts[ order(total_count), cell])
ERCC_mit_Gene_counts[, cell:=factor(cell, levels= sorted_cells )]

ggplot(ERCC_mit_Gene_counts) +
  geom_bar(aes(x=cell, y=count, fill=type), stat="identity", position = "stack") +
  facet_grid(!cell %in% filtered_out ~ .) +
  theme_bw()+
  theme( axis.text.x=element_blank())


ggplot(ERCC_mit_Gene_counts) +
  geom_bar(aes(x=cell, y=count, fill=type), stat="identity", position = "fill") +
  facet_grid(!cell %in% filtered_out ~ .) +
  theme_bw()+
  theme( axis.text.x=element_blank())
```


```{r}

ERCC_mit_Gene_counts[cell %in% filtered_out , filter:="Out" ]
ERCC_mit_Gene_counts[!cell %in% filtered_out , filter:="In" ]


Sup1.a <- ggplot(ERCC_mit_Gene_counts) +
  geom_bar(aes(x=cell, y=count, fill=type), stat="identity", position = "stack") +
  facet_grid(filter ~ .) +
  xlab("Cell")+
  ylab("Read counts")+
  theme_bw()+
  scale_fill_discrete(name = "Count type")+
  theme( axis.text.x=element_blank(),  legend.position = "top")


Sup1.a

```


```{r}

ERCC_mit_Gene_counts[, percentage:= count*100/total_count]

ERCC_mit_Gene_counts_dcast <- dcast(ERCC_mit_Gene_counts[, c("cell",  "type", "percentage")], cell ~ type)


```




# Filtering and TPM normalization



```{r}

Total_cell_counts <- ERCC_mit_Gene_counts$total_count
names(Total_cell_counts) <-  ERCC_mit_Gene_counts$cell





TOTAL_gene_count_melted_filtered <- TOTAL_gene_count_melted[!cell %in%  filtered_out, ]


TOTAL_gene_count_melted_filtered <- merge(TOTAL_gene_count_melted_filtered, gene_lenght, by="Geneid")

TOTAL_gene_count_melted_filtered[, RPK:= count*1e3/Length]
TOTAL_gene_count_melted_filtered[, Total_RPK:=sum(RPK), by=cell]
TOTAL_gene_count_melted_filtered[, TPM:= RPK*1e6/Total_RPK]



TOTAL_gene_count_melted_filtered$total_counts <- Total_cell_counts[TOTAL_gene_count_melted_filtered$cell]
TOTAL_gene_count_melted_filtered[, PM_scaling:=total_counts/1e6]
TOTAL_gene_count_melted_filtered[, RPKM:= count*1e3/(PM_scaling*Length) ]


TOTAL_gene_count_melted_filtered[, log_total_count:=log10(sum(count)), by=cell]


TOTAL_gene_count_filtered <- dcast(TOTAL_gene_count_melted_filtered,  value.var = "RPKM", formula = Geneid ~ cell)

n = ncol(TOTAL_gene_count_filtered)

TOTAL_gene_count_filtered_matrix <- as.matrix(TOTAL_gene_count_filtered)
row.names( TOTAL_gene_count_filtered_matrix) <- TOTAL_gene_count_filtered$Geneid
TOTAL_gene_count_filtered_matrix <- as.matrix(TOTAL_gene_count_filtered_matrix)[, 2:n]
mode(TOTAL_gene_count_filtered_matrix) <- 'numeric'


TOTAL_gene_count_filtered_raw_counts <- dcast(TOTAL_gene_count_melted_filtered,  value.var = "count", formula = Geneid ~ cell)

TOTAL_gene_count_filtered_matrix_raw_counts <- as.matrix(TOTAL_gene_count_filtered_raw_counts)
row.names( TOTAL_gene_count_filtered_matrix_raw_counts) <- TOTAL_gene_count_filtered$Geneid
TOTAL_gene_count_filtered_matrix_raw_counts <- as.matrix(TOTAL_gene_count_filtered_matrix_raw_counts)[, 2:n]
mode(TOTAL_gene_count_filtered_matrix_raw_counts) <- 'numeric'


```


Building the single cell experiment object and performing PCA


```{r, message=FALSE, warning=FALSE}

library(SingleCellExperiment)
library(SC3)
library(scater)

#TOTAL_gene_count_melted_filtered[]

cell_log_counts <- TOTAL_gene_count_melted_filtered[, .(log_total_count=log10(sum(count)) ), by=cell]


filtered_cell_metadata <- merge(metadata, cell_log_counts, by.x="sample_file", by.y="cell")


filtered_cell_metadata <- merge(filtered_cell_metadata, ERCC_mit_Gene_counts_dcast, by.x="sample_file", by.y="cell")

ann <- as.data.frame( filtered_cell_metadata[  ,  c("treatment", "log_total_count", "ERCC", "Mitochondrial")] )

row.names(ann) <- filtered_cell_metadata[  ,  sample_file]


sce <- SingleCellExperiment(
    assays = list(
        counts = TOTAL_gene_count_filtered_matrix,
        logcounts = log2(TOTAL_gene_count_filtered_matrix + 1)
    ), 
    colData = ann
)



# define feature names in feature_symbol column
rowData(sce)$feature_symbol <- rownames(sce)
# remove features with duplicated names
sce <- sce[!duplicated(rowData(sce)$feature_symbol), ]

# define spike-ins
isSpike(sce, "ERCC") <- grepl("ERCC", rowData(sce)$feature_symbol)


scater::plotPCA(sce, colour_by = "treatment", size="log_total_count")

```



Extracting PCA results

```{r}
PCA <- runPCA(sce, colour_by = "treatment", size="log_total_count")
PCA <- as.data.frame(reducedDim(PCA))
PCA$Cell <- row.names(PCA)
PCA <- data.table(PCA)
```


Performing SC3 clustering, results can be interactivelly visuallised with `sc3_interactive(sce)` comand


```{r}
sce <- sc3(sce, ks = 2:3, biology = TRUE)

#sc3_interactive(sce)
```

Incorporating clustering information into the PCA

```{r}

sce_clustering <- as.data.frame(colData(sce))
sce_clustering$Cell <- row.names(sce_clustering)
sce_clustering <- data.table(sce_clustering)

PCA <- merge(sce_clustering, PCA, by="Cell")

sup1.b <- ggplot(PCA) +
  geom_point(aes(PC1, PC2, colour=sc3_2_clusters, shape=treatment, size=log_total_count)) +
  theme_bw() +
  labs(size="Log counts",col="Cluster", shape="Treatment")

sup1.b

```


Ploting SC3 clustering resutls


```{r, fig.height= 7, fig.width=10}


fig5.b <- sc3_plot_consensus(
    sce, k = 2, 
    show_pdata = c(
        "treatment",
        "sc3_2_log2_outlier_score",
        "log_total_count",
        "ERCC",
        "Mitochondrial"
        
    )
)

fig5.b

```



```{r}
sup1.c <- sc3_plot_markers(sce, k = 2)
sup1.c
```


```{r}
sc3_plot_de_genes(sce, k = 2)
```





```{r}

col_data <- as.data.frame(colData(sce))
col_data$cell <- row.names(col_data)
col_data <- data.table(col_data)

cluster1 <- col_data[sc3_2_clusters==1 & sc3_2_log2_outlier_score<=3, ]
cluster2 <- col_data[sc3_2_clusters==2 &  sc3_2_log2_outlier_score<=3, ]

cluster1[, "cell"]
```


# Differential Expession analyses



```{r}
library(monocle)
library(biomaRt)
```



Filtering metadata and getting aditional gene information from biomart

```{r}

metadata_filtered <- metadata[ sample_file %in% colnames(TOTAL_gene_count_filtered), ]
metadata_filtered <- data.frame(metadata_filtered, row.names=metadata_filtered$sample_file)

```


```{r}
gene_table <- data.table(getBM(attributes=c('ensembl_gene_id_version',"ensembl_gene_id", "wikigene_description", "mgi_symbol"),filters = 'ensembl_gene_id_version', values = rownames(TOTAL_gene_count_filtered_matrix) , mart = ensembl))

gene_table <- gene_table[gene_table[ , .I[sample(.N,1)] , by = ensembl_gene_id_version]$V1]  #eliminating duplicates
gene_table <- as.data.frame(gene_table, row.names = gene_table$ensembl_gene_id_version)
gene_table$gene_short_name <- gene_table$mgi_symbol
```


Extracting relevant infromation from gene clusters and differentially expressed genes from SC3

```{r}

row_data <- as.data.frame(rowData(sce))

row_data <- data.table(row_data)


gene_markers <- row_data[sc3_2_markers_auroc>=0.85, ]  

gene_DE <- row_data[sc3_2_de_padj<=0.05,] 


gene_markers <- data.table(merge(gene_table, gene_markers, by.x='ensembl_gene_id_version', by.y='feature_symbol'))

gene_DE <- data.table(merge(gene_table, gene_DE, by.x='ensembl_gene_id_version', by.y='feature_symbol'))

gene_markers[sort(sc3_2_markers_auroc)]

write.table(gene_markers[sort(-sc3_2_markers_auroc)], file = "./output_tables/cluster_marker_info.txt", append = FALSE, quote = F, sep = " ",
            eol = "\n", na = "NA", dec = ".", row.names = F,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")

gene_DE[sort(sc3_2_de_padj)]

write.table( gene_DE , file = "./output_tables/cluster_DE_info.txt", append = FALSE, quote = F, sep = "\t",
            eol = "\n", na = "NA", dec = ".", row.names = F,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")

```


Generating countable for monocle analyses


```{r}

gene_count_monocle <- TOTAL_gene_count_filtered_matrix[gene_table$ensembl_gene_id_version, ]


row.names(gene_table) <- gene_table$ensembl_gene_id_version
rownames(metadata_filtered) <- metadata_filtered$sample_file

```



```{r}


pd <- new("AnnotatedDataFrame", data = metadata_filtered)
fd <- new("AnnotatedDataFrame", data = gene_table)
HSMM <- newCellDataSet(gene_count_monocle, phenoData = pd, featureData = fd, expressionFamily=VGAM::tobit() )

HSMM <- detectGenes(HSMM, min_expr = 0.1)

expressed_genes <- row.names(subset(fData(HSMM),
    num_cells_expressed >= 11))



expressed_genes_mart <- data.table(getBM(attributes=c('ensembl_gene_id_version', "mgi_symbol"),filters = 'ensembl_gene_id_version', values = expressed_genes , mart = ensembl))
```


Attempting to incorporate Spike-ins data for the quantification



```{r}


ERCC_count_table <- TOTAL_gene_count_melted[startsWith(Geneid ,"ERCC"), ]


Gene_lenghts <- gene_lenght$Length
names(Gene_lenghts) <- gene_lenght$Geneid



ERCC_count_table$total_count <-  Total_cell_counts[ERCC_count_table$cell]
ERCC_count_table$lenght <-  Gene_lenghts[ERCC_count_table$Geneid]


ERCC_count_table[, PM_scaling:=total_count/1e6]
ERCC_count_table[, RPKM:= count*1e3/(PM_scaling*lenght) ]


ERCC_count_table[, Geneid:=substring(Geneid, 1, 10)]


ERCC_count_table_matrix <- dcast(ERCC_count_table,  value.var = "RPKM", formula = Geneid ~ cell)

n = ncol(ERCC_count_table_matrix)

ERCC_count_table_matrix <- as.matrix(ERCC_count_table_matrix)
row.names( ERCC_count_table_matrix) <- ERCC_count_table_matrix[,1]
ERCC_count_table_matrix <- as.matrix(ERCC_count_table_matrix)[, 2:n]
mode(ERCC_count_table_matrix) <- 'numeric'




```



```{r}


input.ERCC.annotation<-read.delim("data/ERCC_Controls_Analysis.txt", header=T)
colnames(input.ERCC.annotation)<-c("Resort_ID","ERCC_ID","subgroup","conc_attomoles_ul_Mix1","conc_attomoles_ul_Mix2","exp_fch_ratio","log2_Mix1_Mix2")
rownames(input.ERCC.annotation)<-input.ERCC.annotation[,"ERCC_ID"]




ERCC_Controls_Annotation <- read_delim("data//ERCC_Controls_Annotation.txt", 
    "\t", escape_double = FALSE, trim_ws = TRUE)

ERCC_Controls_Analysis <- read_delim("data//ERCC_Controls_Analysis.txt", 
    "\t", escape_double = FALSE, trim_ws = TRUE)


colnames(ERCC_Controls_Analysis) <- c("sort_ID", "ERCC_ID", "subgroup", "conc_attomoles_ul_Mix1", "conc_attomoles_ul_Mix2", "exp_fch_ratio", "log2(Mix_1/Mix_2)")

ERCC_Controls_Annotation_matrix <- as.matrix(ERCC_Controls_Annotation)

ERCC_Controls_Analysis_matrix <- as.matrix(ERCC_Controls_Analysis)


# Next, use it to estimate RNA counts


rpc_matrix <- relative2abs(HSMM, method = "num_genes")

# Now, make a new CellDataSet using the RNA counts



```

Simulating spike-ins

```{r}

rpc_matrix <- relative2abs(HSMM, method = "num_genes")

HSMM <- newCellDataSet(as(as.matrix(rpc_matrix), "sparseMatrix"),
                phenoData = pd,
                featureData = fd,
                lowerDetectionLimit = 0.5,
                expressionFamily = negbinomial.size())

```


To run the analysis with real Spike-ins, activate the following code. However, the results obtained in this case with simulated spike ins are much better quality and were the ones we used for Gapp et al 2020

```{r}

#rpc_matrix <- relative2abs(HSMM, method = "num_genes", ERCC_controls=ERCC_count_table_matrix[,rownames(metadata_filtered)], ERCC_annotation=input.ERCC.annotation, return_all=F)


#colnames(rpc_matrix) <- rownames(metadata_filtered)
#rownames(rpc_matrix) <- rownames(gene_table)



#HSMM <- newCellDataSet(as(as.matrix(rpc_matrix), "sparseMatrix"),
#                phenoData = pd,
#                featureData = fd,
#                lowerDetectionLimit = 0.5,
#                expressionFamily = negbinomial.size())

```





```{r}

diff_test_res <- differentialGeneTest(HSMM, fullModelFormulaStr = "~treatment", relative_expr = FALSE, cores = 3)


fwrite(diff_test_res, "total.diff.txt", sep="\t" )

```

```{r}
sig_genes <- subset(diff_test_res, qval < 0.05)
sig_genes <- data.table(sig_genes)

sig_genes_expressed_genes <- sig_genes[ensembl_gene_id_version %in% expressed_genes, ]
sig_genes_expressed_genes <- sig_genes_expressed_genes[order(pval ),]
sig_genes_expressed_genes[order(mgi_symbol),c("ensembl_gene_id_version", "gene_short_name", "wikigene_description")]


paste(sig_genes_expressed_genes$gene_short_name, collapse=" ")
```





```{r, fig.height=4, fig.width=15}

HSMM <- detectGenes(HSMM, min_expr = 0.1)

to_be_tested <- row.names(subset(fData(HSMM),
              gene_short_name %in% head(sig_genes$gene_short_name, 10)))








cds_subset <- HSMM[to_be_tested,]

plot_genes_jitter(cds_subset,
                  grouping = "treatment",
                  color_by = "treatment",
                  nrow= 1,
                  ncol = NULL,
                  plot_trend = TRUE)


pData(cds_subset)

cds_subset

```



```{r, fig.height=4, fig.width=15}
to_be_tested <- row.names(subset(fData(HSMM),
              gene_short_name %in% head(sig_genes_total$gene_short_name, 10) ))


cds_subset <- HSMM[to_be_tested,]
plot_genes_jitter(cds_subset,
                  grouping = "treatment",
                  color_by = "treatment",
                  nrow= 1,
                  ncol = NULL,
                  plot_trend = TRUE)



```




# Cluster 1


```{r}

TOTAL_gene_count_filtered_matrix_c1 <- gene_count_monocle[, cluster1[, cell]]


metadata_filtered_c1 <- metadata[sample_file %in% cluster1[, cell], ]
metadata_filtered_c1 <- data.frame(metadata_filtered_c1, row.names=metadata_filtered_c1$sample_file)

rownames(metadata_filtered_c1) <- metadata_filtered_c1$sample_file
rownames(metadata_filtered_c1) <- metadata_filtered_c1$sample_file





```


```{r}
dim(metadata_filtered_c1)
```


```{r}

pd_c1 <- new("AnnotatedDataFrame", data = metadata_filtered_c1)
fd_c1 <- new("AnnotatedDataFrame", data = gene_table)
HSMM_c1 <- newCellDataSet(TOTAL_gene_count_filtered_matrix_c1, phenoData = pd_c1, featureData = fd_c1, expressionFamily=tobit() )


HSMM_c1 <- detectGenes(HSMM_c1, min_expr = 0.1)
#expressed_genes <- row.names(subset(fData(HSMM),
#    num_cells_expressed >= 11))
```


```{r}


fwrite(data.table(expressed_genes_mart$mgi_symbol), file="./expressed_genes.txt", col.names=F)

```


```{r}
rpc_matrix_c1 <- relative2abs(HSMM_c1, method = "num_genes")

HSMM_c1 <- newCellDataSet(as(as.matrix(rpc_matrix_c1), "sparseMatrix"),
                phenoData = pd_c1,
                featureData = fd_c1,
                lowerDetectionLimit = 0.5,
                expressionFamily = negbinomial.size())
```



```{r}

diff_test_res_c1 <- differentialGeneTest(HSMM_c1, fullModelFormulaStr = "~treatment", relative_expr = F)

sig_genes_c1 <- subset(diff_test_res_c1, qval < 0.05)
sig_genes_c1 <- data.table(sig_genes_c1)

fwrite(diff_test_res_c1, "cluster1.diff.txt", sep="\t")

```



# Cluster2


```{r}

TOTAL_gene_count_filtered_matrix_c2 <- gene_count_monocle[, cluster2[, cell]]


metadata_filtered_c2 <- metadata[sample_file %in% cluster2[, cell], ]
metadata_filtered_c2 <- data.frame(metadata_filtered_c2, row.names=metadata_filtered_c2$sample_file)

rownames(metadata_filtered_c2) <- metadata_filtered_c2$sample_file
rownames(metadata_filtered_c2) <- metadata_filtered_c2$sample_file

```


```{r}

pd_c2 <- new("AnnotatedDataFrame", data = metadata_filtered_c2)
fd_c2 <- new("AnnotatedDataFrame", data = gene_table)
HSMM_c2 <- newCellDataSet(TOTAL_gene_count_filtered_matrix_c2, phenoData = pd_c2, featureData = fd_c2, expressionFamily=tobit() )


HSMM_c2 <- detectGenes(HSMM_c2, min_expr = 0.1)
#expressed_genes <- row.names(subset(fData(HSMM),
#    num_cells_expressed >= 11))
```



```{r}
rpc_matrix_c2 <- relative2abs(HSMM_c2, method = "num_genes")

HSMM_c2 <- newCellDataSet(as(as.matrix(rpc_matrix_c2), "sparseMatrix"),
                phenoData = pd_c2,
                featureData = fd_c2,
                lowerDetectionLimit = 0.5,
                expressionFamily = negbinomial.size())
```



```{r}

diff_test_res_c2 <- differentialGeneTest(HSMM_c2, fullModelFormulaStr = "~treatment", relative_expr = F)

sig_genes_c2 <- subset(diff_test_res_c2, qval < 0.05)
sig_genes_c2 <- data.table(sig_genes_c2)

fwrite(diff_test_res_c2, "cluster2.diff.txt", sep="\t" )
```

## Pseudotime

```{r}

HSMM



gene_DE <- data.table(gene_DE)

gene_DE[order(sc3_2_de_padj)][, ensembl_gene_id]



HSMM <- setOrderingFilter(HSMM, gene_DE[order(sc3_2_de_padj)][, ensembl_gene_id])

HSMM3 <- updateCDS(HSMM)


HSMM3 <- estimateSizeFactors(HSMM3)
HSMM3 <- estimateDispersions(HSMM3)

plot_ordering_genes(HSMM3)


data.table(getBM(attributes=c('ensembl_gene_id',"ensembl_transcript_id_version", 'entrezgene', "wikigene_description", "mgi_symbol"),filters = 'ensembl_transcript_id', values = c("ENSMUST00000117661") , mart = ensembl))

cat( gene_DE[, mgi_symbol], sep="\n")

```



```{r}
HSMM <- reduceDimension(HSMM, max_components = 2,
    method = 'DDRTree')

#HSMM <- reduceDimension(my_cds, max_components = 2, num_dim = 5,
#                          reduction_method = 'tSNE', verbose = TRUE)
```


```{r}

HSMM3 <- preprocessCDS(HSMM3)

HSMM3 <- reduceDimension(HSMM3, method = 'tSNE')

HSMM3 <- partitionCells(HSMM3)

HSMM3 <- setOrderingFilter(HSMM3, gene_DE[order(sc3_2_de_padj)][, ensembl_gene_id])


HSMM3 <- learnGraph(HSMM3,  RGE_method = 'SimplePPT',verbose=T)


plot_cell_trajectory(HSMM3)
```



```{r}
HSMM <- orderCells(HSMM)
```




```{r}
paste(gene_DE[, mgi_symbol], sep=" ")

cat( unique(gene_DE[sc3_2_markers_clusts==1, mgi_symbol]), sep=' ')
```


## Intersection 

```{r}

library(limma)


total_sig_matrix <- cbind(as.numeric(diff_test_res$qval<=0.05),
      as.numeric(diff_test_res_c1$qval<=0.05),
      as.numeric(diff_test_res_c2$qval<=0.05)
)


colnames(total_sig_matrix) <- c("Total", "Cluster1", "Cluster2")

row.names(total_sig_matrix) <- diff_test_res$gene_short_name

total_sig_matrix_counts <- vennCounts(total_sig_matrix)

vennDiagram(total_sig_matrix_counts)


names(which(rowSums(total_sig_matrix)==3))



two_clusters_info <-  data.table (diff_test_res[which(rowSums(total_sig_matrix[, 2:3])==1), c("gene_short_name", "wikigene_description", "ensembl_gene_id_version")])


write.table(two_clusters_info, file = "./two_clusters_info.txt", append = FALSE, quote = F, sep = " ",
            eol = "\n", na = "NA", dec = ".", row.names = F,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")


cat(two_clusters_info[, gene_short_name], sep="\n")
```



```{r}
two_clusters_info
```







```{r}

total_sig_table <- data.table(total_sig_matrix)
total_sig_table[, mgi_symbol:=rownames(total_sig_matrix)]
total_sig_table <- total_sig_table[ , c("mgi_symbol", "Total", "Cluster1", "Cluster2")]


total_sig_table[, `:=`(Total=as.logical(Total),
                Cluster1=as.logical(Cluster1),
                Cluster2=as.logical(Cluster2))]


write.table(total_sig_table, file = "./total_sig_table.txt", append = FALSE, quote = F, sep = "\t",
            eol = "\n", na = "NA", dec = ".", row.names = F,
            col.names = TRUE, qmethod = c("escape", "double"),
            fileEncoding = "")
```





```{r}

total_sig_table[ , C1C2:=0]

total_sig_table[ mgi_symbol %in% gene_DE[, mgi_symbol] , C1C2:=1]


total_sig_matrix_C1C2_counts <- vennCounts(total_sig_table[, c(2:5)])

vennDiagram(total_sig_matrix_C1C2_counts)




```


```{r}
total_sig_table[Cluster1>0,]
```


```{r}
library(UpSetR)

diff_intersection <- list(Total = total_sig_table[Total>0, mgi_symbol],
     Cluster1 = total_sig_table[Cluster1>0, mgi_symbol],
     Cluster2 = total_sig_table[Cluster2>0, mgi_symbol])


fig5.c1 <- upset(fromList(diff_intersection), order.by = "freq", nsets = 5, keep.order = T)
fig5.c1     
```


```{r}
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}


```

```{r}


library(eulerr)

diff_intersection <- euler(c("Total" = 58,
                                    "Total&Cluster1" = 19,
                                    "Total&Cluster2" = 19,
                                    "Cluster2" = 15,
                                    "Cluster1" = 8,
                                    "Total&Cluster1&Cluster2" = 7,
                                    "Cluster2&Cluster1" = 4))

                    
    
fig5.c2 <- plot(diff_intersection, font=2, cex=2, alpha=0.5,
     fill=gg_color_hue(3), quantities = list(fontsize = 0))

fig5.c2
```


```{r}
plot_genes_jitter_non_log <- function (cds_subset, grouping = "State", min_expr = NULL, cell_size = 0.75, 
    nrow = NULL, ncol = 1, panel_order = NULL, color_by = NULL, 
    plot_trend = FALSE, label_by_short_name = TRUE, relative_expr = TRUE) 
{
    if (cds_subset@expressionFamily@vfamily %in% c("negbinomial", 
        "negbinomial.size")) {
        integer_expression <- TRUE
    }
    else {
        integer_expression <- FALSE
        relative_expr <- TRUE
    }
    if (integer_expression) {
        cds_exprs <- exprs(cds_subset)
        if (relative_expr) {
            if (is.null(sizeFactors(cds_subset))) {
                stop("Error: to call this function with relative_expr=TRUE, you must call estimateSizeFactors() first")
            }
            cds_exprs <- Matrix::t(Matrix::t(cds_exprs)/sizeFactors(cds_subset))
        }
        cds_exprs <- reshape2::melt(round(as.matrix(cds_exprs)))
    }
    else {
        cds_exprs <- exprs(cds_subset)
        cds_exprs <- reshape2::melt(as.matrix(cds_exprs))
    }
    if (is.null(min_expr)) {
        min_expr <- cds_subset@lowerDetectionLimit
    }
    colnames(cds_exprs) <- c("f_id", "Cell", "expression")
    cds_exprs$expression[cds_exprs$expression < min_expr] <- min_expr
    cds_pData <- pData(cds_subset)
    cds_fData <- fData(cds_subset)
    cds_exprs <- merge(cds_exprs, cds_fData, by.x = "f_id", by.y = "row.names")
    cds_exprs <- merge(cds_exprs, cds_pData, by.x = "Cell", by.y = "row.names")
    cds_exprs$adjusted_expression <- log10(cds_exprs$expression)
    if (label_by_short_name == TRUE) {
        if (is.null(cds_exprs$gene_short_name) == FALSE) {
            cds_exprs$feature_label <- cds_exprs$gene_short_name
            cds_exprs$feature_label[is.na(cds_exprs$feature_label)] <- cds_exprs$f_id
        }
        else {
            cds_exprs$feature_label <- cds_exprs$f_id
        }
    }
    else {
        cds_exprs$feature_label <- cds_exprs$f_id
    }
    if (is.null(panel_order) == FALSE) {
        cds_exprs$feature_label <- factor(cds_exprs$feature_label, 
            levels = panel_order)
    }
    q <- ggplot(aes_string(x = grouping, y = "expression"), data = cds_exprs)
    if (is.null(color_by) == FALSE) {
        q <- q + geom_jitter(aes_string(color = color_by), size = I(cell_size))
    }
    else {
        q <- q + geom_jitter(size = I(cell_size))
    }
    if (plot_trend == TRUE) {
        q <- q + stat_summary(aes_string(color = color_by), fun.data = "mean_cl_boot", 
            size = 0.35)
        q <- q + stat_summary(aes_string(x = grouping, y = "expression", 
            color = color_by, group = color_by), fun.data = "mean_cl_boot", 
            size = 0.35, geom = "line")
    }
    q <- q  + facet_wrap(~feature_label, nrow = nrow, 
        ncol = ncol, scales = "free_y")
    if (min_expr < 1) {
        q <- q + expand_limits(y = c(min_expr, 1))
    }
    q <- q + ylab("Expression") + xlab(grouping) + theme_bw() + theme(legend.position = "top")
    #q <- q + monocle_theme_opts()
    data.table(cds_exprs)
}
```


```{r, fig.height=4, fig.width=10}
library(Hmisc)



HSMM_c1 <- detectGenes(HSMM_c1, min_expr = 0.01)



#to_be_tested_c1 <- row.names(subset(fData(HSMM),
#              gene_short_name %in% total_sig_table[Cluster1 == TRUE , gene]  ))


gene_list <- c("Bcap31", "Hprt", "Rbbp7", "Tcl1")
to_be_tested_c1 <- row.names(subset(fData(HSMM),
              gene_short_name %in% gene_list  ))


cds_subset_c1 <- HSMM_c1[to_be_tested_c1,]

fig5.cluster1 <- plot_genes_jitter_non_log(cds_subset_c1,
                  grouping = "treatment",
                  color_by = "treatment",
                  nrow= 1,
                  ncol = NULL,
                  plot_trend = TRUE,
                  relative_expr = FALSE)



```




```{r,  fig.height=1.5, fig.width=5}
library(Hmisc)

HSMM_c2 <- detectGenes(HSMM_c2, min_expr = 0.01)



#to_be_tested_c2 <- row.names(subset(fData(HSMM),
#              gene_short_name %in% total_sig_table[Cluster2 == TRUE , gene]  ))


gene_list <- c("Bcap31", "Hprt", "Rbbp7", "Tcl1")
to_be_tested_c2 <- row.names(subset(fData(HSMM),
              gene_short_name %in% gene_list  ))





cds_subset_c2 <- HSMM_c2[to_be_tested_c2,]

fig5.cluster2 <- plot_genes_jitter_non_log(cds_subset_c2,
                  grouping = "treatment",
                  color_by = "treatment",
                  nrow= 1,
                  ncol = NULL,
                  plot_trend = TRUE,
                  relative_expr = FALSE)



```






```{r}



diff_test_res_c1.table <- data.table(diff_test_res_c1) #[mgi_symbol %in% unique(fig5.clusters.total$mgi_symbol),  ]
diff_test_res_c2.table <- data.table(diff_test_res_c2) #[mgi_symbol %in% unique(fig5.clusters.total$mgi_symbol),  ]

diff_test_res_c1.table[, Cluster:="1"]
diff_test_res_c2.table[, Cluster:="2"]

diff_test_res.genes.table <- rbind(diff_test_res_c1.table, diff_test_res_c2.table)

diff_test_res.genes.table[ , sig:="NA"]
diff_test_res.genes.table[ qval<=0.05 , sig:="*"]
diff_test_res.genes.table[ qval<=0.01, sig:="**"]
diff_test_res.genes.table[ qval<=0.001, sig:="***"]

diff_test_res.genes.table[, `:=`(start="VEH", end="DEXA", feature_label=mgi_symbol ) ]


```




```{r}

fig5.clusters.total$treatment <- factor(fig5.clusters.total$treatment, levels=c("VEH","DEXA"))

fig5.clusters.total.y_pos <-  fig5.clusters.total[ , .(  y_pos=(max(expression)-min(expression))*0.85 + min(expression)  ), by=c("feature_label", "Cluster")]

diff_test_res.genes.table <- merge(diff_test_res.genes.table[Cluster=="1", ], fig5.clusters.total.y_pos, by=c("feature_label", "Cluster"))


fig5.cluster1$treatment <- factor(fig5.cluster1$treatment , levels=c("VEH", "DEXA"))
        
fig5.f <- ggplot(fig5.cluster1, aes(treatment, expression)) +
  geom_jitter(  aes(colour=treatment) ) +
  stat_summary( aes(colour=treatment), fun.data = "mean_cl_boot",  size = 0.5, shape=4) +
  facet_wrap( . ~ feature_label, scales = "free_y", nrow=1, ) +
  theme_bw() +
  theme(legend.position = "top") +
  geom_signif(data=diff_test_res.genes.table[sig!="NA", ], 
              aes(xmin=start, xmax=end, annotations=sig, y_position=y_pos),
              manual=TRUE)
  
fig5.f  
``` 

```{r, fig.height=30, fig.width=30}






subset_sup.C1 <- HSMM_c1[to_be_tested_sup,]
subset_sup.C2 <- HSMM_c2[to_be_tested_sup,]

sup_jitter.C1 <- plot_genes_jitter_non_log(subset_sup.C1,
                  grouping = "treatment",
                  color_by = "treatment",
                  nrow= 1,
                  ncol = NULL,
                  plot_trend = TRUE,
                  relative_expr = FALSE)


sup_jitter.C2 <- plot_genes_jitter_non_log(subset_sup.C2,
                  grouping = "treatment",
                  color_by = "treatment",
                  nrow= 1,
                  ncol = NULL,
                  plot_trend = TRUE,
                  relative_expr = FALSE)



sup_jitter.C1[, Cluster:="C1" ]
sup_jitter.C2[, Cluster:="C2" ]        
        
sup_jitter.total <- rbind(sup_jitter.C1, sup_jitter.C2)       



sup_jitter.total$Treatment <- factor(sup_jitter.total$treatment, levels=c("VEH","DEXA"))

sup_jitter.total.y_pos <-  sup_jitter.total[ , .(  y_pos=(max(expression)-min(expression))*0.85 + min(expression)  ), by=c("feature_label", "Cluster")]

diff_test_res.genes.table[Cluster=="1", Cluster:="C1"]
diff_test_res.genes.table[Cluster=="2", Cluster:="C2"]

sup_jitter.total.diff_test_res.genes.table <- merge( diff_test_res.genes.table, sup_jitter.total.y_pos, by=c("feature_label", "Cluster"))


sup_jitter.total[feature_label=="", feature_label:=ensembl_gene_id]
sup_jitter.total.diff_test_res.genes.table[feature_label=="", feature_label:=ensembl_gene_id]

remove <- sup_jitter.total.diff_test_res.genes.table[sig=="NA", .N, by="feature_label"][N>1, feature_label]


ggplot(sup_jitter.total[!feature_label %in% remove, ], aes(treatment, expression)) +
  geom_jitter(  aes(colour=treatment) ) +
  stat_summary( aes(colour=treatment), fun.data = "mean_cl_boot",  size = 0.5, shape=4) +
  facet_wrap(  feature_label ~ Cluster , scales = "free_y", ncol=6, ) +
  theme_bw() +
  xlab("Treatment") +
  theme(legend.position = "top") +
  geom_signif(data=sup_jitter.total.diff_test_res.genes.table[sig!="NA" & !feature_label %in% remove , ], 
              aes(xmin=start, xmax=end, annotations=sig, y_position=y_pos),
              manual=TRUE) +
    labs(colour = "Treatment") 
  

```






#### Pesudo-time

###Target Scan ####

```{r}

library(readxl)


get_gene_miRNA_targets <- function(TargetScan_excel){
  
  miR.targets <- read_excel(TargetScan_excel)
  miR.targets.gene_table <- data.table(getBM(attributes=c('ensembl_gene_id_version','ensembl_gene_id', 'ensembl_gene_id_version', 'entrezgene', "wikigene_description", "mgi_symbol"),filters = 'mgi_symbol', values = miR.targets$`Target gene` , mart = ensembl))
  
  miR.targets.gene_table[, mgi_symbol]


}





```






```{r}


miR30.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/list q<0.05/TargetScan7.1__miR-30-5p_384-5p.predicted_targets.xlsx")
miR7210.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/list q<0.05/TargetScan7.1__miR-7210-5p.predicted_targets.xlsx")
miR411.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/list q<0.05/TargetScan7.1__miR-411-5p.predicted_targets.xlsx")
miR7232.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/list q<0.05/TargetScan7.1__miR-7232-3p.predicted_targets.xlsx")

miR150.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/lists q>0.05<0.1/TargetScan7.1__miR-150-5p.predicted_targets.xlsx")
miR7241.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/lists q>0.05<0.1/TargetScan7.1__miR-7241-3p.predicted_targets.xlsx")
miR328.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/lists q>0.05<0.1/TargetScan7.1__miR-328-3p.predicted_targets.xlsx")
miR871.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/lists q>0.05<0.1/TargetScan7.1__miR-871-3p.predicted_targets.xlsx")
miR7233.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/lists q>0.05<0.1/TargetScan7.1__miR-7233-3p.predicted_targets.xlsx")
miR872.names <- get_gene_miRNA_targets("~/Google_Drive/Results/Kathi/TargetScan/lists q>0.05<0.1/TargetScan7.1__miR-872-5p.predicted_targets.xlsx")


```






```{r}
target_len <- c(length(miR30.names),
                length(miR7210.names),
                length(miR411.names), 
                length(miR7232.names), 
                length(miR150.names), 
                length(miR7241.names), 
                length(miR328.names), 
                length(miR871.names), 
                length(miR7233.names), 
                length(miR872.names)
                )

```




```{r}

total_sig_table[, `:=`(
                miR30 = gene %in% miR30.names,
                miR7210= gene %in% miR7210.names,
                miR411= gene %in% miR411.names,
                miR7232= gene %in% miR7232.names,
                miR150= gene %in% miR150.names,
                miR7241= gene %in% miR7241.names,
                miR328= gene %in% miR328.names,
                miR871= gene %in% miR871.names,
                miR7233= gene %in% miR7233.names,
                miR872= gene %in% miR872.names
                )]




```


```{r}

count_matrix <- vennCounts(as.matrix(total_sig_table[, c(2,3,4,5)]))
  
  
  vennCounts(as.matrix(total_sig_table[, c(2,3,4,5)]))



```

```{r}


nrow(total_sig_table[miR30==TRUE, ])

nrow(total_sig_table[Total==TRUE & miR30==TRUE, ])

nrow(total_sig_table[ Total==TRUE ,  ])

nrow(total_sig_table)

s <- length(miR30)
m <- length(miR30)


```



```{r}
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,5)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,6)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,7)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,8)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,9)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,10)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,11)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,12)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,13)])))
vennDiagram(vennCounts(as.matrix(total_sig_table[, c(2,3,4,14)])))

```






```{r}

miR.targets.gene_table[mgi_symbol=="Stim2"]

data.table(getBM(attributes=c('ensembl_gene_id',"ensembl_transcript_id_version", 'entrezgene', "wikigene_description", "mgi_symbol"),filters = 'ensembl_transcript_id', values = c("ENSMUST00000117661") , mart = ensembl))
```



```{r}
listFilters(ensembl)

```










```{r}
library(eulerr)


```



```{r}

names(which(rowSums(total_sig_matrix)==2))

```






```{r}
library("limma")
tstat <- matrix(rt(300,df=10),100,3)
tstat[1:33,] <- tstat[1:33,]+2
clas <- classifyTestsF(tstat,df=10,p.value=0.05)
a <- vennCounts(clas)
print(a)
vennDiagram(a)
```



```{r}
dim(gene_count_monocle)
```




#edgeR

```{r}
L <- list(count = gene_count_monocle, condt = metadata_filtered$treatment)
```


```{r}
dim(gene_count_monocle)
```

```{r}
length(metadata_filtered$treatment)
```


```{r}
suppressPackageStartupMessages(library(edgeR))

run_edgeRQLFDetRate <- function(L) {
  message("edgeRQLFDetRate")
  session_info <- sessionInfo()
  tryCatch({
    timing <- system.time({
      dge <- DGEList(L$count, group = L$condt)
      dge <- calcNormFactors(dge)
      cdr <- scale(colMeans(L$count > 0))
      design <- model.matrix(~ cdr + L$condt)
      dge <- estimateDisp(dge, design = design)
      fit <- glmQLFit(dge, design = design)
      qlf <- glmQLFTest(fit)
      tt <- topTags(qlf, n = Inf)
    })
    
    plotBCV(dge)
    plotQLDisp(fit)
    hist(tt$table$PValue, 50)
    hist(tt$table$FDR, 50)
    limma::plotMDS(dge, col = as.numeric(as.factor(L$condt)), pch = 19)
    plotSmear(qlf,
          de.tags = rownames(tt$table)[which(tt$table$FDR<0.01)])
    
    list(session_info = session_info,
         timing = timing,
         tt = tt,
         summary = data.table(pval = tt$table$PValue,
                        padj = tt$table$FDR,
                        logFC = tt$table$logFC,
                        genes = rownames(tt$table)),
                        qlf=qlf)
  }, error = function(e) {
    "edgeRQLFDetRate results could not be calculated"
    list(session_info = session_info)
  })
}

```


```{r}
raw_counts <- TOTAL_gene_count_filtered_matrix_raw_counts[gene_table$ensembl_gene_id_version, ]

L <- list(count = raw_counts, condt = metadata_filtered$treatment)
res_total <- run_edgeRQLFDetRate(L)
```


```{r}
res_total$qlf

results

rownames(diff_test_res[which(diff_test_res$qval<=0.05),])

plotSmear(res_total$qlf, de.tags = rownames(tt$table)[which(tt$table$FDR<0.1)])



plotSmear(res_total$qlf, de.tags = rownames(diff_test_res[which(diff_test_res$qval<=0.05),]) )  +
abline(v=4,  lty = 2)
```


```{r}
res_total$qlf
```


```{r}



L <- list(count = TOTAL_gene_count_filtered_matrix_c1, condt = metadata_filtered_c1$treatment)
res_c1 <- run_edgeRQLFDetRate(L)

```


```{r}

L <- list(count = TOTAL_gene_count_filtered_matrix_c2, condt = metadata_filtered_c2$treatment)
res_c2 <- run_edgeRQLFDetRate(L)

```



```{r}
res_c1$summary[padj<=0.05,]
```



```{r}
      dge <- DGEList(L$count, group = L$condt)
      dge <- calcNormFactors(dge)
      cdr <- scale(colMeans(L$count > 0))
      design <- model.matrix(~ cdr + L$condt)
      dge <- estimateDisp(dge, design = design)
      fit <- glmQLFit(dge, design = design)
      qlf <- glmQLFTest(fit)
      tt <- topTags(qlf, n = Inf)
```


```{r}
    plotBCV(dge)
    plotQLDisp(fit)
    hist(tt$table$PValue, 50)
    hist(tt$table$FDR, 50)
    limma::plotMDS(dge, col = as.numeric(as.factor(L$condt)), pch = 19)
    plotSmear(qlf)
```

```{r}
plotSmear(qlf,
          de.tags = rownames(tt$table)[which(tt$table$FDR<0.01)])
```


```{r}
df = data.table(pval = tt$table$PValue,
                         padj = tt$table$FDR,
                         gene_names = rownames(tt$table),
                        logFC = tt$table$logFC)
```

```{r}
tt$table
```


```{r}
df[padj<=0.05,]
```

#DEsingle

```{r}
raw_counts <- TOTAL_gene_count_filtered_matrix_raw_counts[gene_table$ensembl_gene_id_version, ]


library(DEsingle)
library(SingleCellExperiment)
data(TestData)

# Convert the test data in DEsingle to SingleCellExperiment data representation
sce <- SingleCellExperiment(assays = list(counts = as.matrix(raw_counts)))

# Specifying the two groups to be compared
# The sample number in group 1 and group 2 is 50 and 100 respectively
#group <- factor(c(rep(1,50), rep(2,100)))

# Detecting the DE genes with SingleCellExperiment input sce
results <- DEsingle(counts = sce, group = factor(metadata_filtered$treatment))

# Dividing the DE genes into 3 categories at threshold of FDR < 0.05
results.classified <- DEtype(results = results, threshold = 0.05)


```



```{r}

raw_counts_c1 <- raw_counts[, cluster1[, cell]]

sce_c1 <- SingleCellExperiment(assays = list(counts = as.matrix(raw_counts_c1)))

results_c1 <- DEsingle(counts = sce_c1, group = factor(metadata_filtered_c1$treatment))

results.classified_c1 <- DEtype(results = results_c1, threshold = 0.05)

```





## MAST


```{r}
suppressPackageStartupMessages({
    library(ggplot2)
    library(GGally)
    library(GSEABase)
    library(limma)
    library(reshape2)
    library(data.table)
    library(knitr)

    library(stringr)
    library(NMF)
    library(rsvd)
    library(RColorBrewer)
    library(MAST)
})
```



```{r}
 metadata[ sample_file %in% cluster1$cell,  ]

metadata[sample_file=="24768_2#060" , ]
```


```{r}

#library(clusterProfiler)
#library(org.Mm.eg.db)
#library(org.Hs.eg.db)
library(biomaRt)
library(data.table)
library(MAST)


ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl")

TOTAL_gene_count_filtered
metadata <- data.table(metadata)

metadata_filtered <- metadata[ sample_file %in% colnames(TOTAL_gene_count_filtered_matrix) & sample_file %in% cluster1$cell, ]
metadata_filtered <- data.frame(metadata_filtered, row.names=metadata_filtered$sample_file)



TOTAL_gene_count_filtered_matrix <- as.matrix(TOTAL_gene_count_filtered[, c("Geneid", cluster1$cell), with=FALSE])
row.names( TOTAL_gene_count_filtered_matrix) <- TOTAL_gene_count_filtered$Geneid
n=ncol(TOTAL_gene_count_filtered_matrix)
TOTAL_gene_count_filtered_matrix <- as.matrix(TOTAL_gene_count_filtered_matrix)[, 2:n]
mode(TOTAL_gene_count_filtered_matrix) <- 'numeric'


TOTAL_gene_count_filtered_matrix_log <- log2(TOTAL_gene_count_filtered_matrix +1)
TOTAL_gene_count_filtered_matrix_log <- TOTAL_gene_count_filtered_matrix_log[which(rowSds(TOTAL_gene_count_filtered_matrix_log)!=0),] 

gene_table <- data.table(getBM(attributes=c('ensembl_gene_id_version','ensembl_gene_id', 'entrezgene', "wikigene_description", "mgi_symbol"),filters = 'ensembl_gene_id_version', values = rownames(TOTAL_gene_count_filtered_matrix_log) , mart = ensembl))

gene_table <- gene_table[gene_table[ , .I[sample(.N,1)] , by = ensembl_gene_id_version]$V1]  #eliminating duplicates

gene_table <- as.data.frame(gene_table, row.names = gene_table$ensembl_gene_id_version)

TOTAL_gene_count_filtered_matrix_log <- TOTAL_gene_count_filtered_matrix_log[gene_table$ensembl_gene_id_version, ]

TOTAL_gene_count_filtered_matrix <- TOTAL_gene_count_filtered_matrix[gene_table$ensembl_gene_id_version, ]

#nzv_cols <- nearZeroVar(TOTAL_gene_count_filtered_matrix_log)
#library("caret")
#if(length(nzv_cols) > 0) TOTAL_gene_count_filtered_matrix_log <- TOTAL_gene_count_filtered_matrix_log[, -nzv_cols]



scaRaw <- FromMatrix(exprsArray=TOTAL_gene_count_filtered_matrix_log, cData=metadata_filtered, fData=gene_table)


```


```{r}
scaRaw
```


```{r}

aheatmap(assay(scaRaw[1:1000,]), labRow='', annCol=as.data.frame(colData(scaRaw)[,c('treatment')]), distfun='spearman')

thres <- thresholdSCRNACountMatrix(assay(sca), nbins = 20, min_per_bin = 30)
par(mfrow=c(5,4))
plot(thres)
```

```{r}
assays(sca) <- list(thresh=thres$counts_threshold, tpm=assay(sca))
expressed_genes <- freq(sca) > freq_expressed
sca <- sca[expressed_genes,]
```





```{r}
sca <- scaRaw

dim(colData(sca))
ncol(TOTAL_gene_count_filtered_matrix_log)

colSums(TOTAL_gene_count_filtered_matrix_log)

```


```{r}

cond<-factor(colData(sca)$treatment)

cond<-relevel(cond,"VEH")
colData(sca)$treatment<-cond



colData(sca)$cngeneson <-  colSums(TOTAL_gene_count_filtered_matrix_log)


zlmCond <- zlm(~treatment + cngeneson, sca)
summaryCond <- summary(zlmCond, doLRT='treatmentDEXA') 


# The following
```


```{r}
freq_expressed <- 0.2
FCTHRESHOLD <- log2(1.5)

summaryDt <- summaryCond$datatable
fcHurdle <- merge(summaryDt[contrast=='treatmentDEXA' & component=='H',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                      summaryDt[contrast=='treatmentDEXA' & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients


hist(log(fcHurdle[, `Pr(>Chisq)`] ))


p.adjust(fcHurdle[, `Pr(>Chisq)`])

table(p.adjust(fcHurdle[, `Pr(>Chisq)`]))


fcHurdle[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]

fcHurdle[fdr<0.5,'fdr']

fcHurdleSig <- merge(fcHurdle[fdr<.05 & abs(coef)>FCTHRESHOLD], as.data.table(mcols(sca)), by='primerid')
setorder(fcHurdleSig, fdr)
```


```{r}
entrez_to_plot <- fcHurdleSig[1:50,primerid]
symbols_to_plot <- fcHurdleSig[1:50,symbolid]
flat_dat <- as(sca[entrez_to_plot,], 'data.table')
ggbase <- ggplot(flat_dat, aes(x=condition, y=thresh,color=condition)) + geom_jitter()+facet_wrap(~symbolid, scale='free_y')+ggtitle("DE Genes in Activated MAIT Cells")
ggbase+geom_violin() 
```




```{r}

TOTAL_gene_count_filtered
metadata

```



## Differential gene expression analysis

```{r}
suppressPackageStartupMessages({
    library(ggplot2)
    library(GGally)
    library(GSEABase)
    library(limma)
    library(reshape2)
    library(data.table)
    library(knitr)
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    library(stringr)
    library(NMF)
    library(rsvd)
    library(RColorBrewer)
    library(MAST)
})

options(mc.cores = detectCores() - 1)

```



```{r}
freq_expressed <- 0.2
FCTHRESHOLD <- log2(1.5)
```

```{r}
data(maits, package='MAST')
dim(maits$expressionmat)
maits$expressionmat
head(maits$cdat)
head(maits$fdat)

scaRaw <- FromMatrix(t(maits$expressionmat), maits$cdat, maits$fdat)


dim(t(maits$expressionmat))
dim(maits$cdat)

View(maits$expressionmat)
View(maits$cdat)
View(maits$fdat)
?FromMatrix
```


```{r}
aheatmap(assay(scaRaw[1:1000,]), labRow='', annCol=as.data.frame(colData(scaRaw)[,c('condition', 'ourfilter')]), distfun='spearman')
```

```{r}
as.data.frame(colData(scaRaw)[,c('treatment')])
metadata
```


```{r}
aheatmap(assay(scaRaw[1:1000,]), labRow='', annCol=as.data.frame(colData(scaRaw)[,c('condition', 'ourfilter')]), distfun='spearman')
```

```{r}
set.seed(123)
plotPCA <- function(sca_obj){
    projection <- rpca(t(assay(sca_obj)), retx=TRUE, k=4)$x
    colnames(projection)=c("PC1","PC2","PC3","PC4")
    pca <- data.table(projection,  as.data.frame(colData(sca_obj)))
    print(ggpairs(pca, columns=c('PC1', 'PC2', 'PC3', 'libSize', 'PercentToHuman', 'nGeneOn', 'exonRate'),
            mapping=aes(color=condition), upper=list(continuous='blank')))
    invisible(pca)
}

plotPCA(scaRaw)

```

```{r}
filterCrit <- with(colData(scaRaw), pastFastqc=="PASS"& exonRate >0.3 & PercentToHuman>0.6 & nGeneOn> 4000)
```




####### Deng et al #####


```{r}



library(SingleCellExperiment)
library(SC3)
library(scater)

deng_reads <- readRDS("./deng-reads.rds" )

sce_deng <- sc3(deng_reads, ks = 8:12, biology = TRUE)


scater::plotPCA(deng_reads, colour_by = "cell_type2")


deng_reads <- sc3_estimate_k(deng_reads)
metadata(deng_reads)$sc3$k_estimation


scater::plotPCA(deng_reads, colour_by = "cell_type1")

table(colData(sce_deng)$cell_type2)




sc3_interactive(sce_deng)
```


```{r}

sc3_plot_consensus(sce_deng, k = 10, show_pdata = "cell_type2")

```


```{r}
sce_deng



deng_reads_early <- deng_reads[ ,  colData(deng_reads)$cell_type1 %in% c("zygote", "2cell", "4cell") ]

as.data.table(colData(deng_reads_early))[ cell_type1=="zygote" , ]


deng_reads_early <- sc3_estimate_k(deng_reads_early)

metadata(deng_reads_early)


deng_reads_early <- sc3(deng_reads_early, ks = 2:7, biology = TRUE)

#sc3_interactive(deng_reads_early)
```


```{r}
sc3_plot_consensus(deng_reads_early, k = 4, show_pdata = c("cell_type1", "cell_type2") )
```



```{r}

library(dplyr)

rownames(deng_reads_early)

rownames(sce)

strsplit(TOTAL_gene_count$Geneid, ".", fixed = TRUE), "[", "", 1)


mgi_symbol_row_names <- getBM(attributes=c('mgi_symbol', "ensembl_gene_id"),filters = 'ensembl_gene_id', values = vapply(strsplit(rownames(sce), ".", fixed = TRUE), "[", "", 1) , mart = ensembl)


sce_rownames <- data.table( vapply(strsplit(rownames(sce), ".", fixed = TRUE), "[", "", 1))

colnames(sce_rownames) <- c("ensembl_gene_id")

sce_rownames <- merge(sce_rownames, mgi_symbol_row_names, all.x=F, by="ensembl_gene_id")
sce_rownames <- sce_rownames[mgi_symbol != "4930553J12Rik", ]


rowData(sce)$ensembl_gene_id <- vapply(strsplit(rownames(sce), ".", fixed = TRUE), "[", "", 1)

sce_mgi_symbol <- sce[ rowData(sce)$ensembl_gene_id %in% sce_rownames$ensembl_gene_id,]

rowData(sce_mgi_symbol)$mgi_symbol <- sce_rownames$mgi_symbol

rowData(sce_mgi_symbol)$feature_symbol <- sce_rownames$mgi_symbol

rownames(sce_mgi_symbol) <- sce_rownames$mgi_symbol





```




```{r}
deng_reads_early
```




```{r}
library(scmap)

deng_reads_early <- selectFeatures(deng_reads_early, suppress_plot = FALSE)



```

```{r}
sce_mgi_symbol <- selectFeatures(sce_mgi_symbol, suppress_plot = FALSE)
```


```{r}
deng_reads_early <- indexCluster(deng_reads_early, cluster_col = "cell_type2")
sce_mgi_symbol <- indexCluster(sce_mgi_symbol , cluster_col = "sc3_2_clusters" )

```


```{r}
heatmap(as.matrix(metadata(deng_reads_early)$scmap_cluster_index))
```

```{r}
library(data.table)

deng_reads_early.metadata <- data.table(as.data.frame(colData(deng_reads_early)))
deng_reads_early.metadata[ , .N , by=c("cell_type1", "cell_type2")]
```


```{r}
cell2_to_time <- scmapCluster(
  projection = sce_mgi_symbol,
  index_list = list(
    muraro = metadata(deng_reads_early)$scmap_cluster_index
  )
)
```


```{r}
deng_reads_early <- indexCluster(deng_reads_early, cluster_col = "sc3_4_clusters")
sce_mgi_symbol <- indexCluster(sce_mgi_symbol , cluster_col = "sc3_2_clusters" )
```



```{r}
heatmap(as.matrix(metadata(deng_reads_early)$scmap_cluster_index))
```


## SCmap


```{r}
cell2_to_time <- scmapCluster(
  projection = sce_mgi_symbol,
  index_list = list(
    muraro = metadata(deng_reads_early)$scmap_cluster_index
  )
)
```





```{r}
plot(getSankey(colData(sce_mgi_symbol)$sc3_2_clusters,  cell2_to_time$scmap_cluster_labs[,1], plot_height=400))
```




```{r}
rowData(sce_mgi_symbol) <- rowData(sce_mgi_symbol$mgi_symbol)
```



```{r}
cell2_to_time

table(colData(sce_mgi_symbol)$sc3_2_clusters, cell2_to_time$scmap_cluster_labs)

cell2_to_time$scmap_cluster_labs

rowData(cell2_to_time)


sce_mgi_symbol

cell2_to_time
```

```{r}
library(ggalluvial)

as.data.frame(cbind(colData(sce_mgi_symbol)$sc3_2_clusters, cell2_to_time$scmap_cluster_labs))

scmap.res <- data.table(data.frame(clusters=colData(sce_mgi_symbol)$sc3_2_clusters, muraro=cell2_to_time$scmap_cluster_labs))

scmap.res.stats <- scmap.res[ , .N, by=c("clusters", "muraro") ]


fig5.c <- ggplot(scmap.res.stats, aes(y=N, axis1=clusters, axis2=muraro)) +
  geom_alluvium(aes(fill = muraro), width = 1/12) +
  geom_stratum(width = 1/12, fill = "white", color = "grey") +
  geom_text(stat = "stratum", infer.label = TRUE) +
  theme(legend.position = "top") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "white"))

fig5.c
  
```

```{r}
fig5.d1+ theme(legend.position = "top")
```

```{r}
fig5.a
```




```{r, fig.height=7, fig.width=10}

plot_grid( plot_grid(Sup1.a, sup1.b, nrow=1, labels = c("A", "B")), as.grob(sup1.c), ncol=1, labels = c("", "C") )

plot_grid()

```





```{r}
library(fields)
x <- c(3,6,8,1,2,2,6,6,7,7,8,8)
y <- c(5,2,3,5,4,6,1,8,3,6,1,7)

df <- data.frame(x,y) 
a  <- c(3,6,8)
b  <- c(5,2,3)
```



```{r}
ss <- function(x) {
  n <- dim(x)[2]
  i <- rep(1:n, n)
  j <- as.vector(t(matrix(i,n)))
  d <- matrix(c(1,1) %*% (x[,i] - x[,j])^2 , n) # The distance matrix entries for `x`
  sum(d[lower.tri(d)])
}
centroid <- function(x) rowMeans(x)
distance2 <- function(x,y) sum((x-y)^2)
#
# Generate two clusters randomly.
#
n.x <- 3; n.y <- 2
x <- matrix(rnorm(2*n.x), 2)
y <- matrix(rnorm(2*n.y), 2)
#
# Compare two formulae.
#
cat("Squared distance between centroids =",
    distance2(centroid(x), centroid(y)),
    "Equivalent value =", 
    (ss(cbind(x,y)) - (n.x + n.y) * (ss(x)/n.x + ss(y)/n.y)) / (n.x*n.y),
    "\n")
```









```{r}


cluster1.VEH.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==1 & treatment=="VEH", Cell]])
cluster1.DEXA.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell]])

cluster2.VEH.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==2 & treatment=="VEH", Cell]])
cluster2.DEXA.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell]])


distance2 <- function(x,y) sum((x-y)^2)

distance2(cluster1.VEH.centroid, cluster1.DEXA.centroid)

```



```{r}
sc3_plot_silhouette(sce, k = 2)
```



```{r}
library(cluster)

cluster1.VEH.pam <- pam(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==1 & treatment=="VEH", Cell]], k=1)
cluster1.DEXA.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell]])

cluster2.VEH.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==2 & treatment=="VEH", Cell]])
cluster2.DEXA.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell]])


```





```{r}

#https://stackoverflow.com/questions/39005958/r-how-to-get-row-column-subscripts-of-matched-elements-from-a-distance-matri

f <- function (i, j, dist_obj) {
  if (!inherits(dist_obj, "dist")) stop("please provide a 'dist' object")
  n <- attr(dist_obj, "Size")
  valid <- (i >= 1) & (j >= 1) & (i > j) & (i <= n) & (j <= n)
  k <- (2 * n - j) * (j - 1) / 2 + (i - j)
  k[!valid] <- NA_real_
  k
  }


## 1D index to 2D index
finv <- function (k, dist_obj) {
  if (!inherits(dist_obj, "dist")) stop("please provide a 'dist' object")
  n <- attr(dist_obj, "Size")
  valid <- (k >= 1) & (k <= n * (n - 1) / 2)
  k_valid <- k[valid]
  j <- rep.int(NA_real_, length(k))
  j[valid] <- floor(((2 * n + 1) - sqrt((2 * n - 1) ^ 2 - 8 * (k_valid - 1))) / 2)
  i <- j + k - (2 * n - j) * (j - 1) / 2
  cbind(i, j)
  }
```


This function gets distance between cells

```{r}

generate_melted.distance.matrix <- function(index_cells, d, d.names){
    
    res <- data.table()
    
        for (c1 in  index_cells){
          
          i = as.numeric(d.names[c1])
          
          for (c2 in index_cells) {
            
            j = as.numeric(d.names[c2])
            
            if (i>j){
            
                ij.dist <- d[f(i , j , d)]
                res <- rbind(res, cbind(c1, c2, i, j, ij.dist))
            }
            
            if (i<j){
            
                ij.dist <- d[f(j , i , d)]

                res <- rbind(res, cbind(c1, c2, i, j, ij.dist))
            
            }                  
            
          }
}

return(res)

}
```




```{r}

generate_melted.inter_cluster_distance.matrix <- function(cluster1, cluster2, d, d.names){
    
    res <- data.table()
    
        for (c1 in  cluster1){

          i = as.numeric(d.names[c1])

          for (c2 in cluster2) {
            
            j = as.numeric(d.names[c2])
            
            if (i>j){
            
                ij.dist <- d[f(i , j , d)]
                res <- rbind(res, cbind(c1, c2, i, j, ij.dist))
            
            }
            
            if (i<j){
            
                ij.dist <- d[f(j , i , d)]
                res <- rbind(res, cbind(c1, c2, i, j, ij.dist))
            
            }            
          }
    }

    return(res)

}
```






```{r}

cluster1.VEH.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==1 & treatment=="VEH", Cell]])
cluster1.DEXA.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell]])

cluster2.VEH.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==2 & treatment=="VEH", Cell]])
cluster2.DEXA.centroid <- rowMeans(TOTAL_gene_count_filtered_matrix_raw_counts[, PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell]])


#d <- dist(as.matrix(PCA[sc3_2_clusters==1 & treatment=="VEH", c("PC1", "PC2")]))
d <- dist(as.matrix(PCA[sc3_2_clusters==1 & treatment=="VEH", c("PC1")]))

#total_PCA.matrix <- dist(as.matrix( PCA[, c('PC1', 'PC2')] ))
total_PCA.matrix <- dist(as.matrix( PCA[, c('PC1')] ))

total_PCA.matrix.names <- seq(1:nrow(PCA))
names(total_PCA.matrix.names) <- PCA[, Cell]


```




```{r}


cluster1.VEH.dist <- generate_melted.distance.matrix(PCA[sc3_2_clusters==1 & treatment=="VEH", Cell],
                                                     total_PCA.matrix,
                                                     total_PCA.matrix.names)
cluster1.DEXA.dist <- generate_melted.distance.matrix(PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell],
                                                      total_PCA.matrix, 
                                                      total_PCA.matrix.names)

cluster2.VEH.dist <- generate_melted.distance.matrix(PCA[sc3_2_clusters==2 & treatment=="VEH", Cell], 
                                                     total_PCA.matrix, 
                                                     total_PCA.matrix.names)
cluster2.DEXA.dist <- generate_melted.distance.matrix(PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell],
                                                      total_PCA.matrix,
                                                      total_PCA.matrix.names)


```




```{r}
cluster1.VEH_DEXA.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==1 & treatment=="VEH", Cell],
                                                                        PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell],
                                                                        total_PCA.matrix,
                                                                        total_PCA.matrix.names)

cluster1.DEXA_VEH.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell],
                                                                        PCA[sc3_2_clusters==1 & treatment=="VEH", Cell],
                                                                        total_PCA.matrix,
                                                                        total_PCA.matrix.names)


cluster2.VEH_DEXA.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==2 & treatment=="VEH", Cell],
                                                                        PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell],
                                                                        total_PCA.matrix,
                                                                        total_PCA.matrix.names)

cluster2.DEXA_VEH.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell],
                                                                        PCA[sc3_2_clusters==2 & treatment=="VEH", Cell],
                                                                        total_PCA.matrix,
                                                                        total_PCA.matrix.names)

```



```{r}

  
  
VEH.cluster1_2.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==1 & treatment=="VEH", Cell],
                                                                     PCA[sc3_2_clusters==2 & treatment=="VEH", Cell],
                                                                     total_PCA.matrix,
                                                                     total_PCA.matrix.names)

VEH.cluster2_1.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==2 & treatment=="VEH", Cell],
                                                                     PCA[sc3_2_clusters==1 & treatment=="VEH", Cell],
                                                                     total_PCA.matrix,
                                                                     total_PCA.matrix.names)



DEXA.cluster1_2.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell],
                                                                     PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell],
                                                                     total_PCA.matrix,
                                                                     total_PCA.matrix.names)

DEXA.cluster2_1.dist <- generate_melted.inter_cluster_distance.matrix(PCA[sc3_2_clusters==2 & treatment=="DEXA", Cell],
                                                                     PCA[sc3_2_clusters==1 & treatment=="DEXA", Cell],
                                                                     total_PCA.matrix,
                                                                     total_PCA.matrix.names)


```





```{r}
VEH.cluster1.mean_dist_int_ext <- merge(cluster1.VEH.dist[ , .(mean_dist_int=(mean(as.numeric(ij.dist)))), by="c1"],
                                        VEH.cluster1_2.dist[ , .(mean_dist_ext=(mean(as.numeric(ij.dist)))), by="c1"] ,
                                        by.x="c1",
                                        by.y="c1")


VEH.cluster2.mean_dist_int_ext <- merge(cluster2.VEH.dist[ , .(mean_dist_int=(mean(as.numeric(ij.dist)))), by="c1"],
                                        VEH.cluster2_1.dist[ , .(mean_dist_ext=(mean(as.numeric(ij.dist)))), by="c1"] ,
                                        by.x="c1",
                                        by.y="c1")


DEXA.cluster1.mean_dist_int_ext <- merge(cluster1.DEXA.dist[ , .(mean_dist_int=(mean(as.numeric(ij.dist)))), by="c1"],
                                        DEXA.cluster1_2.dist[ , .(mean_dist_ext=(mean(as.numeric(ij.dist)))), by="c1"] ,
                                        by.x="c1",
                                        by.y="c1")

DEXA.cluster2.mean_dist_int_ext <- merge(cluster2.DEXA.dist[ , .(mean_dist_int=(mean(as.numeric(ij.dist)))), by="c1"],
                                        DEXA.cluster2_1.dist[ , .(mean_dist_ext=(mean(as.numeric(ij.dist)))), by="c1"] ,
                                        by.x="c1",
                                        by.y="c1")



```





```{r}


centroid <- function(x) rowMeans(x)




PCA_centroids <- PCA[ , .(mean_PC1=mean(PC1), mean_PC2=mean(PC2))  , by=c("treatment", "sc3_2_clusters")]

PCA$treatment <- factor(PCA$treatment , levels = c("VEH", "DEXA"))
PCA_centroids$treatment <- factor(PCA_centroids$treatment , levels = c("VEH", "DEXA"))



fig5.d2 <- ggplot() +
  geom_point(data=PCA, aes(-PC1, PC2, colour=treatment, shape=sc3_2_clusters, size=log_total_count)) +
   geom_point(data=PCA_centroids, aes(-mean_PC1, mean_PC2,  colour=treatment), shape=4, size=5) +
  geom_point(data=PCA_centroids, aes(-mean_PC1, mean_PC2, shape=sc3_2_clusters), size=2) +
  theme_bw() +
  facet_grid( treatment ~ . ) +
  
  labs(size="Log counts",col="Cluster", shape="Treatment")

fig5.d1 <-
  
  ggplot() +
  geom_density(data=PCA,  aes(-PC1, group=treatment, fill=treatment), alpha=0.5, color=NA) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))


```



```{r}

library(ggsignif)

VEH.cluster1.mean_dist_int_ext[, `:=`(cluster="C1", treatment="VEH")]
VEH.cluster2.mean_dist_int_ext[, `:=`(cluster="C2", treatment="VEH")]
DEXA.cluster1.mean_dist_int_ext[, `:=`(cluster="C1", treatment="DEXA")]
DEXA.cluster2.mean_dist_int_ext[, `:=`(cluster="C2", treatment="DEXA")]

TOTAL.silhouette <-  rbind(VEH.cluster1.mean_dist_int_ext,
      VEH.cluster2.mean_dist_int_ext,
      DEXA.cluster1.mean_dist_int_ext,
      DEXA.cluster2.mean_dist_int_ext)


TOTAL.silhouette[  mean_dist_int > mean_dist_ext, silhouette:=(mean_dist_ext - mean_dist_int)/mean_dist_int ]
TOTAL.silhouette[  mean_dist_int < mean_dist_ext, silhouette:=(mean_dist_ext - mean_dist_int)/mean_dist_ext ]


TOTAL.silhouette$treatment <- factor(TOTAL.silhouette$treatment, levels = c("VEH", "DEXA"))


test.c1 <- wilcox.test( TOTAL.silhouette[ cluster=="C1" & treatment=="VEH" , silhouette ],
             TOTAL.silhouette[ cluster=="C1" & treatment=="DEXA" , silhouette ],
             alternative = "less")


test.c2 <- wilcox.test( TOTAL.silhouette[ cluster=="C2" & treatment=="VEH" , silhouette ],
             TOTAL.silhouette[ cluster=="C2" & treatment=="DEXA" , silhouette ],
             alternative = "less")


silhouette.sigtable <- data.table(
start=c("VEH", "VEH"),
end=c("DEXA","DEXA"),
cluster=c("C1","C2"),
p.value=c(test.c1$p.value,test.c2$p.value))


silhouette.sigtable[ , adj.p:=2*p.value ]

silhouette.sigtable[ , sig:="NA"]
silhouette.sigtable[ adj.p<=0.05 , sig:="*"]
silhouette.sigtable[ adj.p<=0.01, sig:="**"]
silhouette.sigtable[ adj.p<=0.001, sig:="***"]



fig5.e <- ggplot(TOTAL.silhouette, aes( x=cluster)) +
  geom_boxplot(aes( y =silhouette, fill = treatment), position="dodge" ) +

  geom_signif(data=silhouette.sigtable,
              aes( y=c(0.95,0.95)),
              y_position=c(0.95, 0.95), xmin=c(0.8, 1.8), xmax=c(1.2, 2.2),
              annotation=silhouette.sigtable$sig ) +

  theme_bw()




```


```{r, fig.height=8, fig.width=7}

fig5.d <- plot_grid(fig5.d1, fig5.d2, ncol = 1, rel_heights = c(0.4,1.5))

fig5.de <- plot_grid(fig5.d, fig5.e, ncol=1, rel_heights = c(1.9,1), labels = c("D", "E"))


fig5.de
```




```{r, fig.height=10, fig.width=15}
library("ggplotify")

fig5.top <- plot_grid(as.grob(fig5.a), fig5.de, nrow=1, rel_widths = c(1.5, 1), labels = c("B", "") )

fig5.bottom <- plot_grid(fig5.c, fig5.f, nrow = 1, rel_widths = c(1, 2), labels = c("C", "F"))

fig5 <- plot_grid(fig5.top, fig5.bottom, rel_heights = c(2, 1), ncol=1)

fig5

```



```{r}

wilcox.test( PCA[sc3_2_clusters==1 & treatment=="VEH", PC1],
             PCA[sc3_2_clusters==1 & treatment=="DEXA", PC1])

wilcox.test( PCA[sc3_2_clusters==2 & treatment=="VEH", PC1],
             PCA[sc3_2_clusters==2 & treatment=="DEXA", PC1])

```

```{r}

wilcox.test( PCA[ sc3_2_clusters==1, ERCC],
             PCA[ sc3_2_clusters==2, ERCC])


wilcox.test( PCA[sc3_2_clusters==1 & treatment=="VEH", ERCC],
             PCA[sc3_2_clusters==1 & treatment=="DEXA", ERCC])


wilcox.test( PCA[sc3_2_clusters==2 & treatment=="VEH", ERCC],
             PCA[sc3_2_clusters==2 & treatment=="DEXA", ERCC])

```



```{r}

wilcox.test( PCA[ sc3_2_clusters==1, log_total_count],
             PCA[ sc3_2_clusters==2, log_total_count],
             alternative="less")


wilcox.test( PCA[sc3_2_clusters==1 & treatment=="VEH", log_total_count],
             PCA[sc3_2_clusters==1 & treatment=="DEXA", log_total_count])


wilcox.test( PCA[sc3_2_clusters==2 & treatment=="VEH", log_total_count],
             PCA[sc3_2_clusters==2 & treatment=="DEXA", log_total_count])

```




```{r}

wilcox.test( PCA[ sc3_2_clusters==1, Mitochondrial],
             PCA[ sc3_2_clusters==2, Mitochondrial],
             alternative="less")


wilcox.test( PCA[sc3_2_clusters==1 & treatment=="VEH", Mitochondrial],
             PCA[sc3_2_clusters==1 & treatment=="DEXA", Mitochondrial])


wilcox.test( PCA[sc3_2_clusters==2 & treatment=="VEH", Mitochondrial],
             PCA[sc3_2_clusters==2 & treatment=="DEXA", Mitochondrial])

```



```{r}
total_sig_table[Cluster1==1, ]
```


```{r}
library(readxl)
Macfarlan_S2 <- data.table(read_excel("~/Google_Drive/Results/Kathi/SuppTable2.xls", 
    skip = 2))


Peaston_S2 <- data.table(read_excel("~/Google_Drive/Results/Kathi/Peaston_S2.xls", 
    skip = 2))

Peaston_S4 <- data.table(read_excel("~/Google_Drive/Results/Kathi/Peaston_S4.xls", 
    skip = 2))



Sharma_S4 <- data.table(read_excel("~/Google_Drive/Results/Kathi/Sharma_S4.xlsx", 
    sheet = "Others"))

Sharma_S4[, Gene:=...2]

Sharma_S4_melt <- melt(Sharma_S4[, c("Gene", "tRF-Gly-GCC_3'LNA", "tRF-Gly-GCC_midLNA_r1", "tRF-Gly-GCC_midLNA_r2")], id="Gene", variable.name="Comparison", value.name="Log2")


Sharma_S4_melt <- merge(Sharma_S4_melt, total_sig_table, by.x="Gene", by.y="mgi_symbol", all.x=T)

Sharma_S4_melt[is.na(Sharma_S4_melt)] <- 0

ggplot(Sharma_S4_melt[Log2>0, ]) +
  geom_boxplot(aes(Comparison, Log2, colour=(Cluster1==1) )) +
  theme_bw()
```

```{r}
library(stringr)

Sharma_S4KG <- data.table(read_excel("~/Google_Drive/Results/Kathi/NIHMS780640-supplement-Table_S4_selected_KG.xlsx"))

Sharma_S4KG.selected <- Sharma_S4KG[ `selected for guille`==TRUE,]


Sharma_S4KG.selected[, Gene:=sub(" .*", "", Gene)]

Sharma_S4KG.selected[ ,source:=sub(" .*", "", `mRNA  Source`)]

Sharma_S4KG.selected[ ,acc:=sub(" .*", "", `mRNA Accession`)] 

Sharma_S4KG.selected[ Gene=="---" & source=="ENSEMBL", acc]



Sharma_genes <-  data.table(getBM(attributes=c('ensembl_transcript_id', "mgi_symbol"),filters = 'ensembl_transcript_id', values = Sharma_S4KG.selected[ Gene=="---" & source=="ENSEMBL", acc] , mart = ensembl))

Sharma_genes[ensembl_transcript_id=="ENSMUST00000111423", mgi_symbol:="AC132444.1"]


Sharma_S4KG.selected <- merge(Sharma_S4KG.selected, Sharma_genes, by.x="acc", by.y="ensembl_transcript_id", all.x=T)


Sharma_S4KG.selected[ Gene=="---" & source=="ENSEMBL" & acc %in% Sharma_genes$ensembl_transcript_id, Gene:=mgi_symbol]

Sharma_S4KG.selected
```

```{r}
total_sig_table[ mgi_symbol %in%  Sharma_S4KG.selected$Gene & Total+Cluster1+Cluster1>0 , ]
```



```{r}
total_sig_table[ mgi_symbol %in%  Macfarlan_S2$gene & Total+Cluster1+Cluster1>0 , ]
```

```{r}
total_sig_table[ mgi_symbol %in%  Peaston_S2$`Gene Symbol` & Total+Cluster1+Cluster1>0 , ]
```

```{r}
total_sig_table[ mgi_symbol %in%  Peaston_S4$`Gene Symbol` & Total+Cluster1+Cluster1>0 , ]
```

```{r}

total_sig_table[ mgi_symbol %in%  Sharma_S4$...2 & Total+Cluster1+Cluster1>0 , ]

total_sig_table[  Total+Cluster1+Cluster1>0,]

```



```{r}
diff_test_res.genes.table
```


```{r}
ggplot(fig5.clusters.total, aes(treatment, expression)) +
  geom_jitter(  aes(colour=treatment) ) +
  stat_summary( aes(colour=treatment), fun.data = "mean_cl_boot",  size = 0.5, shape=4) +
  facet_wrap( Cluster ~ feature_label, scales = "free_y", nrow=2, ) +
  theme_bw() +
  theme(legend.position = "top") +
  geom_signif(data=diff_test_res.genes.table[sig!="NA", ], 
              aes(xmin=start, xmax=end, annotations=sig, y_position=y_pos),
              manual=TRUE)
```






```{r}
# Calculate annotation
anno <- t.test(iris[iris$Petal.Width > 1 & iris$Species == "versicolor", "Sepal.Width"], 
               iris[iris$Species == "virginica", "Sepal.Width"])$p.value

# Make plot with custom x and y position of the bracket
ggplot(iris, aes(x=Species, y=Sepal.Width, fill=Petal.Width > 1)) +
  geom_boxplot(position="dodge") +
  geom_signif(annotation=formatC(anno, digits=1),
              y_position=4.05, xmin=2.2, xmax=3, 
              tip_length = c(0.2, 0.04))

```



```{r}
diff_test_res.genes.table[sig!="NA", ]
```


```{r}

ggplot(fig5.clusters.total, aes(treatment, expression)) +
  geom_jitter(  aes(colour=treatment) ) +
  stat_summary( aes(colour=treatment), fun.data = "mean_cl_boot",  size = 0.5, shape=4) +
  facet_wrap( Cluster ~ feature_label, scales = "free_y", nrow=2, ) +
  theme_bw() +
  theme(legend.position = "top") +
  geom_signif(data=diff_test_res.genes.table[sig!="NA", ], 
              aes(xmin=start, xmax=end, annotations=sig, y_position=y_pos),
              manual=TRUE)
  


```





```{r}
cluster1.DEXA.dist[ , .(mean_dist_int=(mean(as.numeric(ij.dist)))), by="c1"]
```


```{r}
cluster1.DEXA_VEH.dist[ , .(mean_dist_ext=(mean(as.numeric(ij.dist)))), by="c1"]
```


#########


Calculating log2fold changes



```{r}


rpc_matrix_c1.melt <- data.table(reshape2::melt(rpc_matrix_c1))
colnames(rpc_matrix_c1.melt) <- c("gene.ID", "cell", "norm.gene.exp")


metadata_filtered_c1.melted <- data.table( cell=row.names(metadata_filtered_c1), treatment=metadata_filtered_c1$treatment)

rpc_matrix_c1.melt.metatada <- merge(rpc_matrix_c1.melt, metadata_filtered_c1.melted, by="cell")

rpc_matrix_c1.melt.metatada.stats <- rpc_matrix_c1.melt.metatada[ , .(norm.gene.exp.mean=mean(norm.gene.exp), norm.gene.exp.sd=sd(norm.gene.exp))  , by=c("gene.ID", "treatment") ]

rpc_matrix_c1.melt.metatada.stats <- merge(  rpc_matrix_c1.melt.metatada.stats[  treatment=="DEXA",  ], rpc_matrix_c1.melt.metatada.stats[  treatment=="VEH", ], by="gene.ID")

rpc_matrix_c1.melt.metatada.stats[, log2Fold:=log2(norm.gene.exp.mean.x/ norm.gene.exp.mean.y)]

fwrite()


diff_test_res_c1 <- data.table(diff_test_res_c1)



diff_test_res_c1.log2 <- merge(diff_test_res_c1, rpc_matrix_c1.melt.metatada.stats, by.y="gene.ID",  by.x="ensembl_gene_id_version")

diff_test_res_c1.log2[ qval<0.05 &  abs(log2Fold)>1, ]

diff_test_res_c2.log2[ qval<0.05 &  abs(log2Fold)>1, ]

rpc_matrix_c1.melt.metatada.gene <- merge(rpc_matrix_c1.melt.metatada, 
      diff_test_res_c1.log2[ qval<0.05 &  abs(log2Fold)>0.5, c("ensembl_gene_id_version", "gene_short_name"), ] ,
      by.x="gene.ID", 
      by.y="ensembl_gene_id_version")
```


```{r, fig.width=10}

rpc_matrix_c1.melt.metatada.gene$treatment <- factor(rpc_matrix_c1.melt.metatada.gene$treatment, levels=c("VEH", "DEXA"))

ggplot(rpc_matrix_c1.melt.metatada.gene ) +
  geom_jitter(  aes(x=treatment, y=norm.gene.exp, colour=treatment) ) +
  #stat_summary( aes(colour=treatment), fun.data = "mean_cl_boot",  size = 0.5, shape=4) +
  facet_wrap( . ~ gene_short_name, scales = "free_y", nrow=2, ) +
  theme_bw() +
  theme(legend.position = "top")# +
  #geom_signif(data=diff_test_res.genes.table[sig!="NA", ], 
  #            aes(xmin=start, xmax=end, annotations=sig, y_position=y_pos),
  #            manual=TRUE)

```



```{r}
rpc_matrix_c2.melt <- data.table(reshape2::melt(rpc_matrix_c2))
colnames(rpc_matrix_c2.melt) <- c("gene.ID", "cell", "norm.gene.exp")


metadata_filtered_c2.melted <- data.table( cell=row.names(metadata_filtered_c2), treatment=metadata_filtered_c2$treatment)

rpc_matrix_c2.melt.metatada <- merge(rpc_matrix_c2.melt, metadata_filtered_c2.melted, by="cell")

rpc_matrix_c2.melt.metatada.stats <- rpc_matrix_c2.melt.metatada[ , .(norm.gene.exp.mean=mean(norm.gene.exp), norm.gene.exp.sd=sd(norm.gene.exp))  , by=c("gene.ID", "treatment") ]

rpc_matrix_c2.melt.metatada.stats <- merge(  rpc_matrix_c2.melt.metatada.stats[  treatment=="DEXA",  ], rpc_matrix_c2.melt.metatada.stats[  treatment=="VEH", ], by="gene.ID")

rpc_matrix_c2.melt.metatada.stats[, log2Fold:=log2(norm.gene.exp.mean.x/ norm.gene.exp.mean.y)]


diff_test_res_c2.log2 <- data.table(merge(diff_test_res_c2, rpc_matrix_c2.melt.metatada.stats, by.y="gene.ID",  by.x="ensembl_gene_id_version"))

diff_test_res_c1.log2[qval<0.05 &  abs(log2Fold)>1, ]
diff_test_res_c2.log2[qval<0.05 &  abs(log2Fold)>1, ]
```



```{r}
fwrite(diff_test_res_c1.log2, "diff_test_res_c1.log2.txt", sep="\t" )
fwrite(diff_test_res_c2.log2, "diff_test_res_c2.log2.txt", sep="\t" )

```

